app/backend/controllers/portfolio.js

```diff
@@ -0,0 +1,252 @@
+  }).catch(error => {
+    return response.status(400).send(error);
+  });
+}
+  
+/*
+  Get method for portfolios.
+  It returns given portfolio.
+*/
+module.exports.getPortfolio = async (request, response) => {
+  let PortfolioTradingEq = request.models['PortfolioTradingEq']
+  let Portfolio = request.models['Portfolio']
+
+  let portfolio = await Portfolio.find({ _id : request.params['id']})
+  
+  if(portfolio){
+    let tradingEqsObj = await PortfolioTradingEq.find({ PortfolioId : request.params['id']})
+    let tradingEqs = []
+    for (const tradingEq of tradingEqsObj) {
+      tradingEqs.push(tradingEq['TradingEq']);
+    }
+
+    obj = {
+      portfolio: portfolio,
+      tradingEqs: tradingEqs
+    }
+    return response.send(obj)
+  } else {
+    return response.status(400).send({
+      errmsg: "No such portfolio."
+    })
+  }
+}
+
+/*
+  Patch method for editing portfolios.
+*/
+module.exports.renamePortfolio = async (request, response) => {
+  let Portfolio = request.models['Portfolio']
+  const userId = request.session['user']._id
+  const title = request.body["title"];
+  const definition = request.body["definition"];
+  const PortfolioId = request.body['PortfolioId'];
+
+  portfolio = await Portfolio.findOne({ _id : PortfolioId});
+  if(portfolio){
+    Portfolio.updateOne({_id:PortfolioId, userId: userId},{ title: title, definition: definition}) 
+      .then( doc => {
+        return response.status(204).send();
+      }).catch(error => {
+        return response.status(400).send(error);
+      });
+  } else {
+    return response.status(400).send({
+      errmsg: "No such portfolio."
+    })
+  }
+}
+  
+  /*
+    Delete method for portfolios.
+    It deletes given portfolio and portfolio - trading eq. instances.
+  */
+  module.exports.deletePortfolio = async (request, response) => {
+    let Portfolio = request.models['Portfolio']
+    let PortfolioTradingEq = request.models['PortfolioTradingEq']
+  
+    Portfolio.deleteOne({ _id : request.params['id'], userId : request.session['user']._id }, (err, results) => {
+      if(err){
+        return response.status(404).send({
+          errmsg: "Failed to delete portfolio."
+        })
+      }      
+    });
+
+    PortfolioTradingEq.deleteMany({ PortfolioId : request.params['id']}, (err, results) => {
+      if(err){
+        return response.status(404).send({
+          errmsg: "Failed to delete portfolio - trading eq. instances."
+        })
+      }       
+    });
+
+    return response.status(204).send(); 
+  }
+
+  /*
+  Post method for portfolio.
+  It saves portfolio to database.
+*/
+module.exports.addTradingEq = async (request, response) => {
+  let Portfolio = request.models['Portfolio']
+  let PortfolioTradingEq = request.models['PortfolioTradingEq']
+  let tradingEq = request.body['tradingEq']
+  const PortfolioId = request.body['PortfolioId'];
+  // portfolioTradingEq instance to add to the database
+  let portfolioTradingEq = new PortfolioTradingEq({
```

Comment:
what if equipment already exists?

Reply:
return database duplication error, since portfolio id and trading eq form _id which is unique.